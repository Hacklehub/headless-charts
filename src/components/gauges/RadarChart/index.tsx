import React, { useCallback } from 'react';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { line, scaleLinear } from 'd3';
import { select, selectAll } from 'd3-selection';

import { ChartProps } from '../../../types';
import { transition } from 'd3-transition';

export interface RadarChartProps extends ChartProps {
  metrics: { key: string }[];
  classNameMap?: any;
  label: { key: string };
  min?: number;
  max?: number;
}

const RadarChart = ({
  id,
  data,
  className,
  metrics = [],
  label = {
    key: 'name',
  },
  classNameMap = {},
  min = 0,
  max = 1,
  margin = { top: 20, bottom: 20, left: 20, right: 20 },
  drawing = undefined,
}: RadarChartProps) => {
  const refreshChart = useCallback(() => {
    const svg = select(`#${id}`);

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];
    const height = +svg.style('height').split('px')[0];

    const radius =
      Math.min(
        width - margin.left - margin.right,
        height - margin.top - margin.bottom
      ) / 2;

    const g = svg.append('g');
    // .attr('transform', `translate(${width / 2},${height / 2})`);

    const valueFn = scaleLinear().domain([min, max]).range([0, radius]);

    const angleToCoordinate = (angle: number, value: number) => {
      const x = Math.cos(angle) * valueFn(value);
      const y = Math.sin(angle) * valueFn(value);
      return { x: width / 2 + x, y: height / 2 - y };
    };

    const featureData = metrics.map((metric: any, idx: number) => {
      const angle = Math.PI / 2 + (2 * Math.PI * idx) / metrics.length;
      return {
        key: metric.key,
        angle: angle,
        line_coord: angleToCoordinate(angle, max * 0.96),
        label_coord: angleToCoordinate(angle, max * 1.08),
      };
    });

    // draw axis line
    g.selectAll('line')
      .data(featureData)
      .join((enter) =>
        enter
          .append('line')
          .attr('x1', width / 2)
          .attr('y1', height / 2)
          .attr('x2', (d: any) => d.line_coord.x)
          .attr('y2', (d: any) => d.line_coord.y)
          .attr('stroke', 'black')
      );

    // draw axis label
    svg
      .selectAll('.axislabel')
      .data(featureData)
      .join((enter) =>
        enter
          .append('text')
          .attr('class', 'text-xs')
          .attr('x', (d: any) => d.label_coord.x)
          .attr('y', (d: any) => d.label_coord.y)
          .text((d: any) => d.key)
          .attr('text-anchor', 'middle')
      );

    const radarLine: any = line()
      .x((d: any) => d.x)
      .y((d: any) => d.y);

    function getPathCoordinates(d: any) {
      const coordinates = [];
      for (let i = 0; i < metrics.length; i++) {
        const metricName = metrics[i].key;
        const angle = Math.PI / 2 + (2 * Math.PI * i) / metrics.length;
        coordinates.push(angleToCoordinate(angle, d[metricName]));
      }
      coordinates.push(coordinates[0]);
      return coordinates;
    }

    transition();

    const radarArcs = g
      .selectAll('path')
      .data(data)
      .enter()
      .append('path')
      .attr('fill-opacity', 0.7)
      .attr('class', (d: any) =>
        mergeTailwindClasses(
          `stroke-2`,
          d.className,
          classNameMap[d[label.key]]
        )
      )
      .attr('d', (d: any) => {
        if (drawing) {
          const zeroObject = {
            name: d[label.key],
            ...metrics.reduce((acc: any, metric: any) => {
              acc[metric.key] = 0;
              return acc;
            }, {}),
          };

          return radarLine(getPathCoordinates(zeroObject));
        }
        return radarLine(getPathCoordinates(d));
      });

    if (drawing)
      radarArcs
        .transition()
        .duration(drawing?.duration || 1000)
        .delay(drawing.delay || 0)
        .attr('d', (d: any) => {
          return radarLine(getPathCoordinates(d));
        });
  }, [classNameMap, data, id, label, margin, max, metrics, min, drawing]);

  React.useEffect(() => {
    refreshChart();
    return () => {
      selectAll(`#tooltip-${id}`).remove();
    };
  }, [data, refreshChart]);

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(className, defaultChartClassNames)}
    />
  );
};

export default RadarChart;
