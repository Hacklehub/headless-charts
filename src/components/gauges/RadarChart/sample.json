[
  {
    "name": "Arsenal",
    "className": "stroke-red-600 fill-red-600",
    "attack": 90,
    "defense": 80,
    "midfield": 90,
    "goalkeeper": 80,
    "overall": 80
  },
  {
    "name": "Chelsea",
    "className": "stroke-blue-600 fill-blue-600",
    "attack": 80,
    "defense": 80,
    "midfield": 95,
    "goalkeeper": 75,
    "overall": 81
  }
]
