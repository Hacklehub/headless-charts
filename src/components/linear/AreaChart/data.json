[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
