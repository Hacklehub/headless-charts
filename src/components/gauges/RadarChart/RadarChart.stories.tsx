import { Meta, StoryObj } from '@storybook/react';

import RadarChart from '.';
import data from './sample.json';

/**
 * Radar charts are useful when there are five or more variables to compare, along the same value axis (say 0 to 100) across two dimensions.
 *
 * Radar charts are also useful for seeing which variables are scoring high or low within a dataset, making them ideal for displaying performance.
 */
export default {
  title: 'Gauge/RadarChart/Intro',
  component: RadarChart,
  tags: ['autodocs'],
} as Meta;

type Story = StoryObj<typeof RadarChart>;

export const Default: Story = {
  args: {
    id: 'radar-chart',
    data,
    label: {
      key: 'name',
    },
    min: 0,
    max: 100,
    metrics: [
      { key: 'attack' },
      { key: 'defense' },
      { key: 'midfield' },
      { key: 'goalkeeper' },
      { key: 'overall' },
    ],
  },
};

export const Styled: Story = {
  args: {
    ...Default.args,
    id: 'radar-chart-styled',
    classNameMap: {
      Arsenal: 'stroke-red-500 fill-red-500',
      Chelsea: 'stroke-blue-500 fill-blue-500',
    },
  },
};

export const Drawing: Story = {
  args: {
    ...Styled.args,
    id: 'radar-chart-drawing',
    drawing: {
      duration: 1000,
      delay: 1000,
    },
  },
};
