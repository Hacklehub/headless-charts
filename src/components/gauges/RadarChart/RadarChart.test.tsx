import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import RadarChart from '.';
import data from './sample.json';

describe('RadarChart', () => {
  it('renders RadarChart component for snapshot test', () => {
    render(<RadarChart id='radar-chart'
      data={data}
      label={{
        key: 'name',
      }}
      min={0}
      max={100}
      metrics={
        [
          { key: 'attack' },
          { key: 'defense' },
          { key: 'midfield' },
          { key: 'goalkeeper' },
          { key: 'overall' },
        ]} />);
    const radarChart = screen.getByTestId('radar-chart');
    expect(radarChart).toMatchSnapshot();
  });

  it('renders RadarChart component', () => {
    render(<RadarChart id='radar-chart'
      data={data}
      label={{
        key: 'name',
      }}
      min={0}
      max={100}
      metrics={
        [
          { key: 'attack' },
          { key: 'defense' },
          { key: 'midfield' },
          { key: 'goalkeeper' },
          { key: 'overall' },
        ]} />);
    const radarChart = screen.getByTestId('radar-chart');
    expect(radarChart).toBeInTheDocument();
  });

})
