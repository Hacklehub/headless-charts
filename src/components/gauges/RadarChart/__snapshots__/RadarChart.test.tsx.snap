// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadarChart > renders RadarChart component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="radar-chart"
  id="radar-chart"
>
  <g>
    <line
      stroke="black"
      x1="0"
      x2="-1.175660927181459e-15"
      y1="0"
      y2="19.2"
    />
    <line
      stroke="black"
      x1="0"
      x2="18.260285112866946"
      y1="0"
      y2="5.933126291998992"
    />
    <line
      stroke="black"
      x1="0"
      x2="11.285476844015486"
      y1="0"
      y2="-15.533126291998988"
    />
    <line
      stroke="black"
      x1="0"
      x2="-11.285476844015479"
      y1="0"
      y2="-15.533126291998993"
    />
    <line
      stroke="black"
      x1="0"
      x2="-18.26028511286695"
      y1="0"
      y2="5.933126291998986"
    />
    <path
      class="fill-red-600 stroke-red-600"
      d="M0,18L15.217,4.944L10.58,-14.562L-9.405,-12.944L-15.217,4.944L0,18"
      fill-opacity="0.7"
    />
    <path
      class="fill-blue-600 stroke-blue-600"
      d="M0,16L15.217,4.944L11.168,-15.371L-8.817,-12.135L-15.407,5.006L0,16"
      fill-opacity="0.7"
    />
  </g>
  <text
    class="text-xs"
    text-anchor="middle"
    x="-1.3226185430791415e-15"
    y="21.6"
  >
    attack
  </text>
  <text
    class="text-xs"
    text-anchor="middle"
    x="20.542820751975317"
    y="6.674767078498866"
  >
    defense
  </text>
  <text
    class="text-xs"
    text-anchor="middle"
    x="12.696161449517422"
    y="-17.474767078498864"
  >
    midfield
  </text>
  <text
    class="text-xs"
    text-anchor="middle"
    x="-12.696161449517415"
    y="-17.474767078498868"
  >
    goalkeeper
  </text>
  <text
    class="text-xs"
    text-anchor="middle"
    x="-20.54282075197532"
    y="6.674767078498859"
  >
    overall
  </text>
</svg>
`;
